File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws Exception{
		//long start = System.currentTimeMillis();
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		Stack<Integer> st = new Stack<Integer>();
		String input = br.readLine();
		int num = Integer.parseInt(input);
		input = br.readLine();
		String[] arr = input.trim().split(" ");
		List<Integer> seq = new ArrayList<Integer>();
		List<Integer> result = new ArrayList<Integer>();;
		for(int i=0; i<num; i++) {
			seq.add(Integer.parseInt(arr[i]));
			result.add(-1);
		}
		//long end = System.currentTimeMillis();
		//System.out.println("수행시간"+(end-start)+"ms");
		// 데이터 입력 끝
		//start = System.currentTimeMillis();
		
		st.add(0);
		for(int i=1; i<num; i++) {
			int q = st.size();
			for(int j=0; j<q; j++) {
			if(!st.isEmpty() && seq.get(st.peek())<seq.get(i)) {
				result.set(st.peek(), seq.get(i));
				
				st.pop();
				
			}else {
				break;
			}
			}
			st.add(i);
		}
		//end = System.currentTimeMillis();
		//System.out.println("수행시간"+(end-start)+"ms");
		
		//start = System.currentTimeMillis();
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		for(int i=0; i<num; i++) {
				bw.write(result.get(i)+" ");
			
		}
		bw.flush();
		//end = System.currentTimeMillis();
		//System.out.println("수행시간"+(end-start)+"ms");
	
		

		

	}

}

File: 백준/Gold/17298. 오큰수/오큰수.py
N = int(input())
lst = list(map(int,input().split()))
index_lst= []
result_lst=[-1]*N
stk = []
stk.append([lst[0],0])# value, index
for i in range(1, N):
    index = i
    # print(stk)
    if not stk :
        stk.append([lst[i],i])
    else:
        if stk[-1][0]<lst[i]:
            while True:
                if index !=0 and stk and stk[-1][0]<lst[i]:
                    temp_lst=stk.pop()
                    # print("temp_lst",temp_lst)
                    result_lst[temp_lst[1]]=lst[i]

                    # print(i, index, "실행되었습니다")
                else:
                    stk.append([lst[i],i])
                    break

        else:
            stk.append([lst[i],i])

    index_lst.append([index, i, lst[i]])

# print(index_lst)

# for i in index_lst:
#     for j in range(i[0], i[1]):
#         if result_lst[j]==-1:
#             result_lst[j]=i[2]

for i in result_lst:
    print(i, end= " ")
